odl/trafos/backends/pynfft_bindings.py

```diff
@@ -9,13 +9,96 @@
 """
+
+import numpy as np
+
 try:
     import pynfft
     PYNFFT_AVAILABLE = True
 except ImportError:
     PYNFFT_AVAILABLE = False
 
-__all__ = ('PYNFFT_AVAILABLE', )
+__all__ = ('PYNFFT_AVAILABLE', 'normalize_samples')
+
+
+def normalize_samples(samples, grid_spacing, out=None):
+    r"""Normalize samples to [-0.5; 0.5[.
+
+    Parameters
+    ----------
+    samples : array-like
+        Array of unnormalized samples. Its shape is expected to be
+        ``(M, d)``, where ``M`` is the number of samples, and ``d`` the
+        dimension of the space.
+    grid_spacing : positive float or sequence of positive float
+        Step between grid points in the space on which the non-uniform Fourier
+        transform is applied. A sequence must have length ``d``. A single
+        float is interpreted as the same spacing in each dimension.
+    out : numpy.ndarray, optional
+        Array in which the result should be stored. Must have the same shape
+        as ``samples``.
+
+    Returns
+    -------
+    norm_samples : numpy.ndarray
+        Normalized samples. See Notes for details. If ``out`` was given, the
+        returned array is a reference to it.
+
+    Notes
+    -----
+    For a uniformly discretized space with grid spacing
+    :math:`s > 0 \in \mathbb{R}^d`, samples :math:`\xi_k \in \mathbb{R}^d`
+    are normalized to
+
+    .. math::
+        \bar \xi_k = \frac{s \odot \xi_k}{2\pi}.
+
+    Since the reciprocal grid extends from :math:`-\pi s^{-1}` to
+    :math:`\pi s^{-1}`, this transformation normalized the samples to lie
+    between :math:`-0.5` and :math:`0.5`. Due to implementation details,
+    the value :math:`0.5` has to be replaced with :math:`-0.5`, which gives
+    the same result for periodic inputs.
+    """
+    samples = np.array(samples, copy=False, ndmin=2)
+    if samples.ndim != 2:
+        raise ValueError(
+            '`samples` must have 2 dimensions, got array with `ndim={}`'
+            ''.format(samples.ndim)
+        )
+    ndim = samples.shape[1]
+
+    grid_spacing_in = grid_spacing
+    try:
+        iter(grid_spacing)
+    except TypeError:
+        grid_spacing = [float(grid_spacing)] * ndim
+    else:
+        grid_spacing = [float(s) for s in grid_spacing]
+
+    if len(grid_spacing) != ndim:
+        raise ValueError(
+            '`grid_spacing` must have length of `samples.shape[1]`, but '
+            '{} != {}'.format(len(grid_spacing), ndim)
+        )
+
+    if any(s <= 0 for s in grid_spacing):
+        raise ValueError(
+            '`grid_spacing` must be positive, got {}'.format(grid_spacing_in)
+        )
+
+    if out is not None and out.shape != samples.shape:
+        raise ValueError(
+            '`out.shape` must be equal to `samples.shape`, but {} != {}'
+            ''.format(out.shape, samples.shape)
+        )
+
+    grid_spacing = np.array(grid_spacing, dtype=samples.dtype)
+    norm_samples = np.multiply(
+        samples, (grid_spacing[None, :] / (2 * np.pi)), out=out
```

Comment:
I don't understand how this is going to normalize the samples. I think overall I have a lack of understanding of the whole normalization issue. But if I take my current use case (greatly simplified), this is what I would get:
```python
grid_spacing = 1
samples = np.array([[80, 80], [80, 79]])
norm_samples = normalize_samples(samples, grid_spacing)
```

And here `norm_samples` doesn't have values in `[-0.5; 0.5[`. Maybe you could tell me what you think about this specific case.

Reply:
Those samples would be too large in magnitude, so you're right, they wouldn't be normalized.

The point is, the relation between grid spacing and Fourier space samples is not arbitrary. Rather, a signal in real space with a grid spacing of `s` is [band-limited](https://en.wikipedia.org/wiki/Bandlimiting) and only contains frequencies up to `pi / s` in magnitude (the extra factor `2 * pi` comes from varying definitions of Fourier transforms).
Or, putting it the other way around, for a function given on a grid with spacing `s`, one can only obtain Fourier transform samples in a cube from `-pi / s` to `+pi / s`, and points outside are invalid.
By normalizing with `s / (2 * pi)` those boundaries are mapped to the `[-0.5, 0.5]` bounds of NFFT.

Reply:
Ok I understand. I guess it's a matter of convention and how you want to see the griding in your frequency space. In my use case, I think the easiest way is to look at it in terms of largest and smallest value and then assume the grid spacing is going to be right, whereas in your case you look at it in terms of grid spacing and then assume the largest and smallest value is going to be right.

If you think that it's easier to relate it to the uniform FT this way I think it's a fair point. I would say that in my use case we lose a bit of user-friendliness, but I don't know if this is true overall for all the usages. Anyway since this is a tradeoff and we are not really sure about the benefit of my pov, let's go with yours which has a set one.

Thanks for taking the time to explain.

Reply:
What I was mostly worried about is to use a normalization that assumes that the most extreme samples define your min/max frequencies. If you decide to only sample in a subset of that space, there would be no way to do that. The new min/max would be the ones deciding how to scale, and the scaling for all others would change. In a nutshell, I think the normalization of the samples should not be determined by the samples themselves.

Now, in your case, it may be safe to assume that you always sample up to theoretical maximum and in a symmetrical way around the origin. If you're worried about user-friendliness for that specific case, maybe it helps to make a little helper function that converts between the two conventions. For instance, calculate a grid spacing from the samples such that your normalization and mine are equivalent.

Reply:
> What I was mostly worried about was using a normalization that assumes that the most extreme samples define your min/max frequencies. If you decide to only sample in a subset of that space, there would be no way to do that.

Yes totally agree that it would be fatal. Defining the max and min was indeed necessary (but in this case was adding an extra argument, which I had called `space`).